### PR TITLE
Main

### PR DESCRIPTION
**Pull request by Hroy02**

### CODE REVIEW

File: src/ServerHandler.cpp [hunk 1]
```diff
@@ -129,25 +139,53 @@ void ServerHandler::parseMessageFrame(Frame messageFrame) {
...
+    while(_connectionHandler == nullptr || !_connectionHandler->getFrameAscii(frameString, STOMP_DELIMITER.c_str()[0])) {
+        if(_shouldTerminate) return;
+    }
     Frame frame = Frame(frameString);
     FrameCommand cmd = frame.getCommand();
+    if(!_loggedIn && cmd == RECEIPT) {
+        delete _connectionHandler;
+        _connectionHandler = nullptr;
+        sendFrameToClient(frame);
+    }
```
Comment (IdanTo1) [inline src/ServerHandler.cpp:151]: why is that?

File: src/UserHandler.cpp [hunk 2]
```diff
@@ -139,6 +142,9 @@ void UserHandler::addBook(std::string& line) {
     split(line, cmdParams, CMD_DELIMITER);
     std::string genre = cmdParams[1];
     std::string book = cmdParams[2];
+    for(int i = 3; i < cmdParams.size(); i++) {
+        book += cmdParams[i];
+    }
```
Comment (IdanTo1) [inline src/UserHandler.cpp]: nice catch

File: src/UserHandler.cpp [hunk 3]
```diff
@@ -192,11 +198,18 @@ void UserHandler::logout(std::string& line) {
     if (ansCmd == ERROR) {
         std::cout << ans.getBody() << std::endl;
     }
+    else {// ansCmd == RECEIPT
+        _inventory.clear();
+        while(!_queues.framesFromServer.empty()) _queues.framesFromServer.pop();
```
Comment (IdanTo1) [inline src/UserHandler.cpp:204]: :laughing: 
I asked myself if it needed and decided to pass. 
**sync on the queue is required...**

File: src/UserHandler.cpp [hunk 4]
```diff
@@ -192,11 +198,18 @@ void UserHandler::logout(std::string& line) {
...
+        std::cout << "Disconnected from server" << std::endl;
+    }
+
 }
 
 void UserHandler::parseLine(std::string line) {
     if (line == "bye") {
         _shouldTerminate = true;
+        sendFrame(Frame(_actionFrameCommandMap[LOGOUT]));
```
Comment (IdanTo1) [inline src/UserHandler.cpp]: didn't we say we give it up?

**Review by IdanTo1 — CHANGES_REQUESTED** [review]
looks very good.
some minor fixes
**make sure to review it as well, as most of the code here is mine**

File: src/ServerHandler.cpp (hunk 1 above)
Comment (Hroy02) [reply src/ServerHandler.cpp:151]: We want to get the receipt for the disconnect before actually disconnecting

File: src/UserHandler.cpp (hunk 4 above)
Comment (Hroy02) [reply src/UserHandler.cpp]: Right, will change

File: src/UserHandler.cpp (hunk 3 above)
Comment (Hroy02) [reply src/UserHandler.cpp:204]: Good idea, will add!

File: src/ServerHandler.cpp (hunk 1 above)
Comment (IdanTo1) [reply src/ServerHandler.cpp:151]: ok, now i saw where it is. good job

File: src/UserHandler.cpp (hunk 3 above)
Comment (IdanTo1) [reply src/UserHandler.cpp:204]: I think you should use lock_guard. I'm not perfectly sure whats the different, but i rather not wait to find out

**Review by IdanTo1 — CHANGES_REQUESTED** [review]
only one little thing.
change it and i think we're good to go

**Review by IdanTo1 — APPROVED** [review]

**Pull request merged**